src/relay/transforms/to_a_normal_form.cc

```diff
@@ -223,6 +223,17 @@ class Fill : ExprFunctor<Expr(const Expr&, const Var&)>, private transform::Lexi
     bool not_included = include_set_ && include_set_->find(orig) == include_set_->end();
     if (!v.defined() && not_included) {
       return annotated_expr;
+    } else if (const LetNode* let = AsIgnoringOnDevice<LetNode>(now)) {
+      // Instead of making a nested binding "let var = (let x = ...; bindings...; body)", we push
+      // the inner bindings into the outer scope and bind body to var, giving
+      // "let x = ...; bindings...; let var = body;" as the resulting bindings.
+      Expr e = GetRef<Expr>(let);
+      while (const LetNode* inner_let = AsIgnoringOnDevice<LetNode>(e)) {
+        GetScope(orig)->let_list->Push(inner_let->var, inner_let->value);
```

Comment:
I believe (and again, sorry this is awkward -- this is why Lily is chipping away at virtual_device_) you'll need a MaybeOnDevice on the inner_let->value. Basically, we have to preserve any device information on the let-bound values, you can't unwrap then ignore.

I'd not be at all surprised if no tests pick this up. In an ideal world every device aware pass would have unit tests exercising it's handling of devices.

Reply:
I *think* the inner values should already be wrapped, right? assuming they were already handled by a recursive `Compound` call

Reply:
Shoot you're write, I misread the loop sorry!